### PR TITLE
Add E2E test for PnP

### PR DESCRIPTION
**Pull request by hihigupt**

### CODE REVIEW

File: tests/e2e/device/e2e_device_commands.c [hunk 1]
```diff
@@ -780,6 +785,150 @@ static uint32_t prvE2ETestVerifyDesiredPropertiesCommandExecute( E2E_TEST_COMMAN
 }
 /*-----------------------------------------------------------*/
 
+/**
+ * Execute Writeable property response command
```
Comment (danewalton) [inline tests/e2e/device/e2e_device_commands.c:789]: This is more of a nit I think with this whole file (including the file name I guess). The `_commands` used in various places now has a different connotation with PnP being a thing. Could we look to change this to something else: maybe `_task` or `_job`? When I read this and the subsequent function call, it seems to be mixing properties and commands.

File: tests/e2e/device/e2e_device_commands.c [hunk 2]
```diff
@@ -780,6 +785,150 @@ static uint32_t prvE2ETestVerifyDesiredPropertiesCommandExecute( E2E_TEST_COMMAN
...
+    uint16_t usTelemetryPacketID;
+    uint32_t ulRequestId;
+    az_span xJsonSpan = az_span_create( ( uint8_t * ) xCMD->pulReceivedData, xCMD->ulReceivedDataLength );
+    az_span xKey = az_span_create( ucScratchBuffer, e2etestMESSAGE_BUFFER_SIZE / 3 );
+    az_span xValue = az_span_create( ucScratchBuffer + e2etestMESSAGE_BUFFER_SIZE / 3, e2etestMESSAGE_BUFFER_SIZE / 3 );
+    az_span xValueReceived = az_span_create( ucScratchBuffer + 2 * e2etestMESSAGE_BUFFER_SIZE / 3, e2etestMESSAGE_BUFFER_SIZE / 3 );
+    AzureIoTJSONWriter_t xJSONWriter;
+
+    ulStatus = prvGetValueForKey( xJsonSpan, e2etestDESIRED_PROPERTY_KEY, &xKey );
+    RETURN_IF_FAILED( ulStatus, "failed to find desired_property_key!" );
```
Comment (danewalton) [inline tests/e2e/device/e2e_device_commands.c:808]: nit: here and 811 for consistency
```suggestion
    RETURN_IF_FAILED( ulStatus, "Failed to find desired_property_key!" );

```

File: tests/e2e/device/e2e_device_commands.c [hunk 3]
```diff
@@ -780,6 +785,150 @@ static uint32_t prvE2ETestVerifyDesiredPropertiesCommandExecute( E2E_TEST_COMMAN
...
+                                                       ucScratchBuffer2,
+                                                       ( uint32_t ) AzureIoTJSONWriter_GetBytesUsed( &xJSONWriter ),
+                                                       &ulRequestId ) != eAzureIoTSuccess )
+    {
+        RETURN_IF_FAILED( e2etestE2E_TEST_FAILED, "Failed to send reported properties" );
+    }
+    else if( AzureIoTHubClient_ProcessLoop( pxAzureIoTHubClient,
+                                            e2etestPROCESS_LOOP_WAIT_TIMEOUT_IN_SEC ) != eAzureIoTSuccess )
+    {
+        RETURN_IF_FAILED( e2etestE2E_TEST_FAILED, "recevice timeout" );
```
Comment (danewalton) [inline tests/e2e/device/e2e_device_commands.c:840]:
```suggestion
        RETURN_IF_FAILED( e2etestE2E_TEST_FAILED, "Test failed: receive timeout" );

```

File: tests/e2e/device/e2e_device_commands.c [hunk 4]
```diff
@@ -780,6 +785,150 @@ static uint32_t prvE2ETestVerifyDesiredPropertiesCommandExecute( E2E_TEST_COMMAN
...
```
Comment (danewalton) [inline tests/e2e/device/e2e_device_commands.c:914]:
```suggestion
        RETURN_IF_FAILED( e2etestE2E_TEST_FAILED, "Test failed: receive timeout" );

```

File: tests/e2e/device/e2e_device_commands.c [hunk 5]
```diff
@@ -780,6 +785,150 @@ static uint32_t prvE2ETestVerifyDesiredPropertiesCommandExecute( E2E_TEST_COMMAN
...
+    uint16_t usTelemetryPacketID;
+    uint32_t ulRequestId;
+    az_span xJsonSpan = az_span_create( ( uint8_t * ) xCMD->pulReceivedData, xCMD->ulReceivedDataLength );
+    az_span xKey = az_span_create( ucScratchBuffer, e2etestMESSAGE_BUFFER_SIZE / 3 );
+    az_span xValue = az_span_create( ucScratchBuffer + e2etestMESSAGE_BUFFER_SIZE / 3, e2etestMESSAGE_BUFFER_SIZE / 3 );
+    az_span xComponent = az_span_create( ucScratchBuffer + 2 * e2etestMESSAGE_BUFFER_SIZE / 3, e2etestMESSAGE_BUFFER_SIZE / 3 );
+    AzureIoTJSONWriter_t xJSONWriter;
+
+    ulStatus = prvGetValueForKey( xJsonSpan, e2etestCOMPONENT_KEY, &xComponent );
+    RETURN_IF_FAILED( ulStatus, "failed to find component!" );
```
Comment (danewalton) [inline tests/e2e/device/e2e_device_commands.c:878]: Same here can just do a find all and replace
```suggestion
    RETURN_IF_FAILED( ulStatus, "Failed to find component!" );
```

File: tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts [hunk 6]
```diff
@@ -215,4 +216,34 @@ describe("mainTest", () => {
         let expectedMessage = await iothubTestCore.runTestCommand(testHubConnectionString, testDeviceInfo.deviceId, get_twin_cmd);
         let rc = await iothubTestCore.verifyTelemetryMessage(testHubConnectionString, testDeviceInfo.deviceId, expectedMessage);
     })
+
+    it("Device Writable Properties Response", async function() {
+        const payload = { "temprature" : (Math.floor(Math.random() * 10000000000).toString())};
```
Comment (danewalton) [inline tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts:221]:
```suggestion
        const payload = { "temperature" : (Math.floor(Math.random() * 10000000000).toString())};

```

Comment (danewalton) [reply tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts:221]: and elsewhere

File: tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts [hunk 7]
```diff
@@ -215,4 +216,34 @@ describe("mainTest", () => {
...
+        const reported_properties_cmd = new e2eTestCommands.CommandTestData_ReportComponentPropertiesCommands("thermostate", "temprature", payload["temprature"]);
+        let expectedMessage = await iothubTestCore.runTestCommand(testHubConnectionString, testDeviceInfo.deviceId, reported_properties_cmd);
+        let rc = await iothubTestCore.verifyTelemetryMessage(testHubConnectionString, testDeviceInfo.deviceId, expectedMessage);
+        let twin = await iothubTestCore.getTwinProperties(testHubConnectionString, testDeviceInfo.deviceId);
+        console.log(twin)
+        var twinObject = JSON.parse(twin);
+        assert.ok(twinObject.reported.thermostate);
+        assert.ok(twinObject.reported.thermostate.__t);
+        assert.ok(twinObject.reported.thermostate.temprature);
+        assert.strictEqual(twinObject.reported.thermostate.temprature, payload["temprature"]);
```
Comment (danewalton) [inline tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts:247]:
```suggestion
        assert.strictEqual(twinObject.reported.thermostate.temprature, payload["temperature"]);

```

Comment (danewalton) [reply tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts:247]: can do a find all and replace here

File: tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts [hunk 8]
```diff
@@ -215,4 +216,34 @@ describe("mainTest", () => {
...
+
+    it("Device Writable Properties Response", async function() {
+        const payload = { "temprature" : (Math.floor(Math.random() * 10000000000).toString())};
+        const writable_properties_cmd = new e2eTestCommands.CommandTestData_WritablePropertiesResponseCommands("temprature", payload["temprature"]);
+        let expectedMessage = await iothubTestCore.runTestCommand(testHubConnectionString, testDeviceInfo.deviceId, writable_properties_cmd);
+        let rc = await iothubTestCore.verifyTelemetryMessage(testHubConnectionString, testDeviceInfo.deviceId, expectedMessage);
+        let twin = await iothubTestCore.getTwinProperties(testHubConnectionString, testDeviceInfo.deviceId);
+        console.log(twin)
+        var twinObject = JSON.parse(twin);
+        assert.ok(twinObject.reported.temprature);
```
Comment (danewalton) [inline tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts:228]: is this (and elsewhere) just asserting that this value exists?

**Review by danewalton — APPROVED** [review]

File: tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts (hunk 8 above)
Comment (hihigupt) [reply tests/e2e/service/iothub_client/e2e_iot_hub_client_test.ts:228]: yes, this tells this object exists and the value is tested line 230and 233.

**Comment by hihigupt** [comment]
Test failure because of issue fixed in PR: https://github.com/Azure/azure-iot-middleware-freertos/pull/166

File: tests/e2e/device/e2e_device_commands.c (hunk 1 above)
Comment (hihigupt) [reply tests/e2e/device/e2e_device_commands.c:789]: This change will increase the scope of this PR.. I think we should take care of this in separate PR.

Comment (danewalton) [reply tests/e2e/device/e2e_device_commands.c:789]: yup definitely didn't want it part of this one

**Pull request merged**